engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java

```diff
@@ -0,0 +1,268 @@
+                remove(client, containerId);
+                return new ExecutionOutcome(Stage.COMPLETED, null, "");
+            } catch (IOException e) {
+                log.error("Execution [{}] failed:", analysis.getId(), e);
+                callbackWithStdOut.accept(Stage.EXECUTE, e.getMessage());
+                return new ExecutionOutcome(Stage.COMPLETED, e.getMessage(), "");
+            }
+        }, executor);
+        return new SQLService.SqlOverseer(analysis.getId(), started, stdout, future);
+    }
+
+    private void runContainer(DockerClient dockerClient, String containerId, File analysisDir, AnalysisSyncRequestDTO analysis, BiConsumer<String, String> callback) {
+        String execFileName = analysis.getExecutableFileName();
+
+        log.info("Starting container: {}", containerId);
+        dockerClient.startContainerCmd(containerId).exec();
+
+        String codeDir = "/bin";
+        log.info("Copying study into container: {}", containerId);
+        dockerClient.copyArchiveToContainerCmd(containerId)
+                .withHostResource(analysisDir.getPath())
+                .withRemotePath(codeDir).exec();
+
+        String baseDir = codeDir + "/" + analysisDir.getName() + "/";
+        String workingDir = baseDir + (execFileName.split("/").length > 1 ? execFileName.split("/")[0] : "");
+        log.info("Working dir {}", workingDir);
+        String rscriptPath = baseDir + execFileName;
+        log.info("Script path {}", rscriptPath);
+        String runCommandId = dockerClient.execCreateCmd(containerId)
+                .withAttachStdout(true)
+                .withAttachStderr(true)
+                .withCmd("Rscript", rscriptPath)
+                .withWorkingDir(workingDir).exec()
+                .getId();
+
+        ResultCallback.Adapter<Frame> adapter = getAdapter(analysis, callback);
+        try {
+            log.info("Running script: {}", rscriptPath);
+            dockerClient.execStartCmd(runCommandId).exec(adapter).awaitCompletion(runtimeTimeOutSec, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            log.error("InterruptedException occurred whilst running {} in {}, message: {}", rscriptPath, containerId, e.getMessage());
+            callback.accept(Stage.EXECUTE, e.getMessage());
+        } finally {
+            stop(dockerClient, containerId);
+        }
+    }
+
+    private ResultCallback.Adapter<Frame> getAdapter(AnalysisSyncRequestDTO analysis, BiConsumer<String, String> callback) {
+        return new ResultCallback.Adapter<Frame>() {
+            @Override
+            public void onNext(Frame item) {
+                log.info(item.toString(), analysis);
+                try {
+                    String output = new String(item.getPayload());
+                    log.info("Execution output: {}", output);
+                    callback.accept(Stage.EXECUTE, output);
+                } catch (Exception e) {
```

Comment:
What kind of exception you are expecting here?

Reply:
It would be `RuntimeException`.